src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1673,13 +1669,34 @@ RouterJob(Query *originalQuery, PlannerRestrictionContext *plannerRestrictionCon
 
 	/* check if this query requires master evaluation */
 	bool requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
+	FastPathRestrictionContext *fastPathRestrictionContext =
+		plannerRestrictionContext->fastPathRestrictionContext;
+
+	/*
+	 * We prefer to defer shard pruning/task generation to the
+	 * execution when the parameter on the distribution key
+	 * cannot be resolved (or no paramater exists).
+	 */
+	if (fastPathRestrictionContext->fastPathRouterQuery &&
```

Comment:
One idea I had was to make this decision more explicit by adding  `fastPathRestrictionContext -> distKeyEqualsParam` , and checking for it here.

Any preference?

Reply:
I've done that, we can revert the commit `Defer pruning only for parameters` to see the previous implementation